### ISSUE TITLE
有单机多卡模式么？ 

### ISSUE THREAD

**Issue by Fighting-JJ** [Sierkinhane/crnn_chinese_characters_rec#53]
up主，想过修改成可以在多个显卡上运行么？
或者可以提供点参考资料。

单单使用Dataparalle，在CTCloss那块会出现batchsize不匹配的问题；
这个batchsize是在dataloader那里就成了para里设定值（B）的一半，还是取完所有B后，split成B/2的？

pytorch的API太高层了，不知道底层怎么实现的，思维有点混乱。不如tensorflow的单机多卡流程清晰，数据分成n份，feed给各个模型，然后各自计算loss与需要变化的grad，最后将grad平均后，同步到多卡。

**Comment by JingLiJJ** [comment]
> up主，想过修改成可以在多个显卡上运行么？
> 或者可以提供点参考资料。
> 
> 单单使用Dataparalle，在CTCloss那块会出现batchsize不匹配的问题；
> 这个batchsize是在dataloader那里就成了para里设定值（B）的一半，还是取完所有B后，split成B/2的？
> 
> pytorch的API太高层了，不知道底层怎么实现的，思维有点混乱。不如tensorflow的单机多卡流程清晰，数据分成n份，feed给各个模型，然后各自计算loss与需要变化的grad，最后将grad平均后，同步到多卡。

你好。请问你解决这个问题了嘛？

**Comment by Fighting-JJ** [comment]
没有

**Comment by Lanme** [comment]
我也遇到了这个问题，输入(batch,imgH,imgW)是(1024,32,280)单卡的preds.size()是torch.Size([71, 1024, 6009])，但是多卡的是torch.Size([142, 512, 6009])，不过我看网上说是在前向传播的时候才等分的，怎么输出序列的大小也变了呢？

**Comment by leigaoxiang** [comment]
请问你这个问题解决了？
@Lanme

**Comment by Donaghys** [comment]
> 我也遇到了这个问题，输入(batch,imgH,imgW)是(1024,32,280)单卡的preds.size()是torch.Size([71, 1024, 6009])，但是多卡的是torch.Size([142, 512, 6009])，不过我看网上说是在前向传播的时候才等分的，怎么输出序列的大小也变了呢？

请问你解决这个问题了吗

**Comment by wting861006** [comment]
用pytorch-lightning 不会有这个问题，我修改了一下训练代码，发出来给大家参考一下
`

    import argparse
    import torch
    import torch.backends.cudnn as cudnn
    from torch.utils.data import DataLoader, random_split
    import pytorch_lightning as pl
    from easydict import EasyDict as edict
    import yaml
    import lib.models.crnn as crnn
    import lib.utils.utils as utils
    from lib.dataset import get_dataset
    import lib.config.alphabets as alphabets
    from lib.utils.utils import model_info
    from tensorboardX import SummaryWriter
    
    
    def parse_arg():
        parser = argparse.ArgumentParser(description="train crnn")
    
        parser.add_argument('--cfg', help='experiment configuration filename', required=True, type=str)
    
        args = parser.parse_args()
        with open(args.cfg, 'r') as f:
            # config = yaml.load(f, Loader=yaml.FullLoader)
            config = yaml.safe_load(f)
            config = edict(config)
    
        config.DATASET.ALPHABETS = alphabets.alphabet
        config.MODEL.NUM_CLASSES = len(config.DATASET.ALPHABETS)
    
        return config
    
    
    
    class CRNN_TRAIN(pl.LightningModule):
        def __init__(self,train_dataset,val_dataset,converter,config):
            super(CRNN_TRAIN,self).__init__()
            self.net=model
            self.criterion=criterion
            self.optimizer=optimizer
            self.train_dataset=train_dataset
            self.val_dataset=val_dataset
            self.converter=converter
            self.config=config
            self.n_correct = 0
    
        def forward(self,x):
            return self.net(x)
    
        def training_step(self,batch,batch_idx ):
            inp, idx=batch
            labels = utils.get_batch_label(self.train_dataset, idx)
            preds = self(inp)
            batch_size = inp.size(0)
            text, length = self.converter.encode(labels)
            preds_size = torch.IntTensor([preds.size(0)] * batch_size)
            loss=self.criterion(preds,text,preds_size,length)
            return loss
    
        def validation_step(self,batch,batch_idx):
            inp, idx=batch
            labels = utils.get_batch_label(self.val_dataset, idx)
            preds = self(inp)
            batch_size = inp.size(0)
            text, length = self.converter.encode(labels)
            preds_size = torch.IntTensor([preds.size(0)] * batch_size)
            loss = self.criterion(preds, text, preds_size, length)
            _, preds = preds.max(2)
            preds = preds.transpose(1, 0).contiguous().view(-1)
            sim_preds = self.converter.decode(preds.data, preds_size.data, raw=False)
            for pred, target in zip(sim_preds, labels):
                if pred == target:
                    self.n_correct += 1
            num_test_sample = self.config.TEST.NUM_TEST_BATCH * self.config.TEST.BATCH_SIZE_PER_GPU
            if num_test_sample > len(self.val_dataset):
                num_test_sample = len(self.val_dataset)
            accuracy = self.n_correct / float(num_test_sample)
            self.log("val_loss", loss, prog_bar=True)
            self.log("val_acc", accuracy, prog_bar=True)
    
        def configure_optimizers(self):
            return self.optimizer
    
    if __name__=="__main__":
        config = parse_arg()
        cudnn.benchmark = config.CUDNN.BENCHMARK
        cudnn.deterministic = config.CUDNN.DETERMINISTIC
        cudnn.enabled = config.CUDNN.ENABLED
    
        output_dict = utils.create_log_folder(config, phase='train')
    
        writer_dict = {
            'writer': SummaryWriter(log_dir=output_dict['tb_dir']),
            'train_global_steps': 0,
            'valid_global_steps': 0,
        }
    
        model = crnn.get_crnn(config)
        criterion = torch.nn.CTCLoss()
        last_epoch = config.TRAIN.BEGIN_EPOCH
        optimizer = utils.get_optimizer(config, model)
    
        if isinstance(config.TRAIN.LR_STEP, list):
            lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(
                optimizer, config.TRAIN.LR_STEP,
                config.TRAIN.LR_FACTOR, last_epoch - 1
            )
        else:
            lr_scheduler = torch.optim.lr_scheduler.StepLR(
                optimizer, config.TRAIN.LR_STEP,
                config.TRAIN.LR_FACTOR, last_epoch - 1
            )
    
        if config.TRAIN.FINETUNE.IS_FINETUNE:
            model_state_file = config.TRAIN.FINETUNE.FINETUNE_CHECKPOINIT
            if model_state_file == '':
                print(" => no checkpoint found")
    
            checkpoint = torch.load(model_state_file, map_location='cpu')
            if 'state_dict' in checkpoint.keys():
                checkpoint = checkpoint['state_dict']
    
            from collections import OrderedDict
    
            model_dict = OrderedDict()
            for k, v in checkpoint.items():
                if 'cnn' in k:
                    model_dict[k[4:]] = v
            model.cnn.load_state_dict(model_dict)
            if config.TRAIN.FINETUNE.FREEZE:
                for p in model.cnn.parameters():
                    p.requires_grad = False
    
        elif config.TRAIN.RESUME.IS_RESUME:
            model_state_file = config.TRAIN.RESUME.FILE
            if model_state_file == '':
                print(" => no checkpoint found")
            print("load model:", model_state_file)
            checkpoint = torch.load(model_state_file, map_location='cpu')
            if 'state_dict' in checkpoint.keys():
                model.load_state_dict(checkpoint['state_dict'])
                last_epoch = checkpoint['epoch']
                # optimizer.load_state_dict(checkpoint['optimizer'])
                # lr_scheduler.load_state_dict(checkpoint['lr_scheduler'])
            else:
                model.load_state_dict(checkpoint)
    
        train_dataset = get_dataset(config)(config, is_train=True)
        train_loader = DataLoader(
            dataset=train_dataset,
            batch_size=config.TRAIN.BATCH_SIZE_PER_GPU,
            shuffle=config.TRAIN.SHUFFLE,
            num_workers=config.WORKERS,
            pin_memory=config.PIN_MEMORY,
        )
        val_dataset = get_dataset(config)(config, is_train=False)
        val_loader = DataLoader(
            dataset=val_dataset,
            batch_size=config.TEST.BATCH_SIZE_PER_GPU,
            shuffle=config.TEST.SHUFFLE,
            num_workers=config.WORKERS,
            pin_memory=config.PIN_MEMORY,
        )
        converter = utils.strLabelConverter(config.DATASET.ALPHABETS)
    
        model=CRNN_TRAIN(train_dataset,val_dataset,converter,config)
        # trainer = pl.Trainer(gpus=3, precision=32, max_epochs=config.TRAIN.END_EPOCH,auto_lr_find=True)
        trainer = pl.Trainer(devices=3, precision=32, 
        max_epochs=config.TRAIN.END_EPOCH,auto_lr_find=True,accelerator="gpu",strategy="ddp")
        trainer.fit(model, train_dataloaders=train_loader, val_dataloaders=val_loader)
`

**Comment by sunjunlishi** [comment]
@Lanme 我的问题和你一样，不知道怎么解决

torch.Size([92, 3, 48, 600])
torch.Size([150, 46, 101])
torch.Size([1620])
torch.Size([92])
torch.Size([92])


model(inp).*********************1
model(inp).*********************2
text.size()
torch.Size([92, 3, 48, 600])
torch.Size([75, 92, 101])
torch.Size([1618])
torch.Size([92])
torch.Size([92])
model(inp).*********************1
model(inp).*********************2
text.size()
torch.Size([92, 3, 48, 600])
torch.Size([75, 92, 101])
torch.Size([1622])
torch.Size([92])
torch.Size([92])

**Comment by sunjunlishi** [comment]
@Lanme 
简单的尝试解决，更改维度：
        preds = model(inp).cpu()
        
        ps = preds.size()
        
        new_shape = ((int)(ps[0]/2), ps[1]*2, ps[2])
        preds = preds.view(*new_shape)
